doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,414 @@
+
+## Easy Cases: Fully Symbolic or Fully Evaluated
+
+Before digging into the more advanced functionality for custom functions, we
+should mention two common cases that are much simpler, the case where the
+function is fully symbolic, and the case where the function is fully
+evaluated.
+
+(custom-functions-fully-symbolic)=
+### Fully Symbolic
+
+If your function `f` has no mathematical properties you
+want to define on it, and should never evaluate on any arguments, you can
+create an undefined function using `Function('f')`
+
+```
+>>> from sympy import symbols, Function
+>>> x = symbols('x')
+>>> f = Function('f')
+>>> f(x)
+f(x)
+>>> f(0)
+f(0)
+```
+
+This is useful, for instance, when solving [ODEs](ode-docs).
+
+This is also useful if you only wish to create a symbol that depends on
+another symbol for the purposes of differentiation. By default, SymPy assumes
+all symbols are independent of one another:
+
+```
+>>> from sympy.abc import x, y
+>>> y.diff(x)
+0
+```
+
+To make a symbol depend on another symbol, you should make it a function that
+explicitly depends on that symbol.
+
+```
+>>> y = Function('y')
+>>> y(x).diff(x)
+Derivative(y(x), x)
+```
+
+If you want your function to have additional behavior, for example, to have a
+custom derivative, or to evaluate on certain arguments, you should create a
+custom `Function` subclass as described below. However, undefined functions do
+support one additional feature, which is that assumptions can be defined on
+them, using the same syntax as used by symbols. This defines the assumptions
+of the output of the function, not the input (that is, it defines the
+function's range, not its domain).
+
+```
+>>> g = Function('g', real=True)
+>>> g(x)
+g(x)
+>>> g(x).is_real
+True
+```
+
+To make a function's assumptions depend on its input in some way, you should
+create a custom `Function` subclass as described below.
+
+(custom-functions-fully-evaluated)=
+### Fully Evaluated
+
+At the other end of the spectrum are functions that always evaluate to
+something no matter what their inputs are. These functions are never left in
+an unevaluated, symbolic form like `f(x)`.
+
+In this case, you should just use a normal Python function using the `def`
+keyword.
+
+```
+>>> def f(x):
+...     if x == 0:
+...         return 0
+...     else:
+...         return x + 1
+>>> f(0)
+0
+>>> f(1)
+2
+>>> f(x)
+x + 1
+```
+
+If you find yourself defining an `eval` method on a `Function` subclass where
+you always return a value and never return `None`, you should consider just
+using a normal Python function instead, as there is no benefit to using a
+symbolic `Function` subclass in that case.
+
+Note that in many cases, functions like these can be represented directly
+using SymPy classes. For example, the above function can be represented
+symbolically using `Piecewise`. The function can then be evaluated using
+{meth}`subs() <sympy.core.basic.Basic.subs>`.
```

Comment:
Also perhaps worth mentioning `Lambda` and maybe even `lambdify` here:
```python
In [19]: f = Lambda(x, Piecewise((0, Eq(x, 0)), (x+1, True)))

In [20]: f
Out[20]: 
    ⎧  0    for x = 0
x ↦ ⎨                
    ⎩x + 1  otherwise

In [21]: f(0)
Out[21]: 0
```

Reply:
Mentioning lambdify is a good idea. I thought about Lambda, but I'm a little unsure because I'm not sure just how well supported it is throughout SymPy. Very few things make use of it.

Reply:
The potential advantage of `Lambda` is that you get a symbolic object that can be used like a function with Python's call syntax (assuming that's what is wanted):
```python
In [1]: eq_y = Lambda(x, Piecewise((0, Eq(x, y)), (1, True)))

In [2]: eq_y
Out[2]: 
    ⎧0  for x = y
x ↦ ⎨            
    ⎩1  otherwise

In [3]: eq_y(y+1)
Out[3]: 1

In [4]: eq_y(y)
Out[4]: 0
```
Unlike other options the `Lambda` function can be used with `subs` to substitute something that is not the "argument" of the function:
```python
In [5]: eq_4 = eq_y.subs(y, 4)

In [6]: eq_4
Out[6]: 
    ⎧0  for x = 4
x ↦ ⎨            
    ⎩1  otherwise
```

Reply:
The thing I don't understand about `Lambda` is that it's an `Expr`. But does it actually make sense to include a `Lambda` inside of other expressions? You can't call it unless the top-level expression is a `Lambda`. Does something like `Lambda(...) + 1` even make sense?

Reply:
I would prefer it if `Lambda` wasn't a subclass of Expr. There is a backwards way of calling it:
```python
In [1]: f = Lambda(x, x**2)

In [2]: f
Out[2]: 
     2
x ↦ x 

In [3]: exp(f)
Out[3]: 
      2
 x ↦ x 
ℯ      

In [4]: exp(f).rcall(3)
Out[4]: 
 9
ℯ
```
I'm not sure I've ever seen that used in the wild though.